interpreter/exec/simd.ml

```diff
@@ -315,10 +331,46 @@ struct
       let num_lanes = lanes F64x2
     end)
 
+  let clamp x low high = min (max x low) high
+
+  (* Narrow two v128 into one v128 by using to_shape on both operands,
+   * concatenating them, clamping the element to between low and high,
+   * then of_shape to reconstruct a v128. *)
+  let narrow to_shape of_shape low high x y =
+    let xy = (to_shape x) @ (to_shape y) in
+    let narrow i = Int32.(clamp i (of_int low) (of_int high)) in
+    of_shape (List.map narrow xy)
+
+  module I8x16_convert = struct
+    let narrow_s = narrow Rep.to_i16x8 Rep.of_i8x16 (-128) 127
+    let narrow_u = narrow Rep.to_i16x8 Rep.of_i8x16 0 255
+  end
+
+  module I16x8_convert = struct
+    let narrow_s = narrow Rep.to_i32x4 Rep.of_i16x8 (-32768) 32767
+    let narrow_u = narrow Rep.to_i32x4 Rep.of_i16x8 0 65535
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xff))
+    let widen take_or_drop f x =
+      Rep.of_i16x8 (List.map f ((take_or_drop 8) (Rep.to_i8x16 x)))
+    let widen_low_s = widen (Lib.List.take) (fun x -> x)
+    let widen_high_s = widen (Lib.List.drop) (fun x -> x)
+    let widen_low_u = widen (Lib.List.take) unsigned_mask
+    let widen_high_u = widen (Lib.List.drop) unsigned_mask
```

Comment:
You could inline this as
```suggestion
    let widen_low_u = widen Lib.List.take (Int32.logand 0xffl)
    let widen_high_u = widen Lib.List.drop (Int32.logand 0xffl)
```

Or even simpler than using a function parameter: just have `widen` take a mask and pass in either 0xffffffffl or 0xffl.

Reply:
Sg, made that change.